Move indentation services down to Workspace layer.

CyrusNajmabadi opened the pull request:
For reasons that have never been quite clear to me, our indentation services were defined at the EditorFeatures layer (maybe because they thought it was an editor concept?).  This was especially odd as similar features (like Formatting) are defined at the Workspaces layer.

This change simply moves things down to the Workspace layer (while also keeping around and obsoleting the existing editor-interfaces since they are consumed by F#/TS/JS).  

This makes thigns much cleaner, and makes it possible to push other features down (like the wrapping-code-actions).

CyrusNajmabadi commented:
Tagging @dpoeschl @sharwell

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Indentation/IIndentationService.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <see langword="null"/> if the line in question is not blank and thus indentation should
+        /// be deferred to the formatting command handler to handle.
+        /// </summary>
+        IndentationResult? GetDesiredIndentation(Document document, int lineNumber, CancellationToken cancellationToken);
```
Note: i still hate this (and do not understand why the API id designed this way).  I would really like to make this non-nullable, and move anything that the "formatting command handler" would do into this logic.   It seems super bizarre that we have this strange split instead of making the formatting command handler *not* handle 'enter' and having that just be smart idnent, which goes through this.  

Unfortunately, this side of formatting/indentation is just an utter mess and really hard to detween.

CyrusNajmabadi replied:
I'm going to try to make this change as a followup.  But it would be nice to get this in first to keep the size of the PRs smaller.

CyrusNajmabadi replied:
@heejaechang Can you shed some light here on the design.  There seem to be three different systems here, which all are stateful and which try to infer what the other would be up to.

First, there is ISmartIndent.  This is the system the editor calls in on us to ask us "where should we be indented to".  This seems like the most natural location to place any indentation logic/computation.

However, we seem to have *two other* systems that are external to this.  First, there is the SmartTokenFormatHandler.  This handles enter, and sometimes processes where the enter shoudl go.  However, it's not *always* the case that enter is handled here.  Sometimes this instead says "will FormatCommandHandler handle enter?" and then defers to *that* system to compute where enter can go.

It's really not clear here why there is so much split in responsibilities in terms of computing this.  Nor is it clear when each is supposed to be figuring this out.  I'd really like to move this all down to the IIndentationService.  First, as far as i can tell, this should always succeed (so we can get rid of an unnecessary nullable).  We should always be able to figure out indentation and the API should reflect taht.  Second, it would be nice to have a *single* service that any downstream consumer can use to ask "what should the indentation be here".  Right now, afaict, that's (nearly) impossible to ask since you would have to ask three separate systems "are *you* responsible for figuring out indentation?  and if so, what do you think the indentation should be?"   This seems highly error prone and confusing.

Can you shed some light here on why the design is this way, and how we might be able to remove this complexity and just have a single standard mechanism for doing this?

heejaechang replied:
I dont know what it is doing. I don't know who wrote this ISynchronousIndentationService, IBlankLineIndentationService. or IndentationResult.

what I wrote, I believe, was someting return virtual position on the blank line for editor API. 

hmm.. looking at what we have now (http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/SmartIndent/IIndentationService.cs,a4f4522e8b8eb434,references)

not even sure why the API is async. I guess someone just extract out things from editor API but didn't properly refactor it? ah.. no one in roslyn implement that service...

okay, I remember this ShouldUseSmartTokenFormatterInsteadOfIndenter, but loos like it is like moved from its original context and copied to this place... anyway, this is something that see whether we can use formatting rule to figure out indentation or not. if it can, it should use the rules so that indentation doesn't need to re-implement all those logic to figure out what is right indentation from formattiong options such as case should be indented or not, or brace should be indented or not and etc. list of all options is in formattion option page.

this indenter and abstract indenter code seems have my initial code but looks like moved around. but not actually properly refactored but just moved around.

and smart token formatter is the thing that contains all the heuristic to figure out whether what span it should call formatting to or whether it should just move caret to right indentation and etc.

type char handling could be moved from formatting command handler to this, but looks like formatting command handler is done in a way it can handle anything that is not related to indentation, but smart token command handler is written to handle case where 2 can get mixed.

anyway, I am not sure why IEditorFormattingService is introduced and why everything is refactored this way. I remember code reviewing this and gave some feedbacks on refactor it differently but I guess my feedback wasn't taken.

hmmm smart token formatter used to handle all cases such as format range, and etc, not sure why ISmartTokenFormatter only do FormatToken and FormatRange is not part of it and etc.

I guess it is time to clean things up here.

CyrusNajmabadi replied:
Note: a later PR removes this.

CyrusNajmabadi commented:
@dotnet/roslyn-infrastructure @jasonmalinowski @jaredpar 

Do you know what causes this error on pipelines: 

```
 IlAsmDeploy -> F:\vsagent\1\s\artifacts\bin\IlAsmDeploy\Debug\netcoreapp2.1\win-x64\IlAsmDeploy.dll
F:\vsagent\1\s\src\Tools\ILAsm\IlAsmDeploy.csproj(46,11): error MSB4006: There is a circular dependency in the target dependency graph involving target "_PublishILAsm".
```

?

CyrusNajmabadi commented:
Oh, the formatting stuff is even weirder than i first though.  It looks like we actually run SmartTokenCommandHandler and FormatCommandHandler in the same chain.  Though there's nothing that indicates that these should be ordered in any particular fashion.  This is definitely a bit strange as that means you could hand Format handle enter first, followed by SmartToken, or vice versa.

In my current debug-session, SmartTokenCommandHandler called into FormatCommandHandler.  There FormatCommandHandler bailed out very quickly (it looks like it only tries to handle close-parens and using-statements??).  We then got back to SmartTokenCommandHandler, which did a bunch of checks, and then bailed out because it thought the editor handled things... 

It's all pretty confusing about who/what/where and when is responsible for handling thse things.  @heejaechang can you shed some light on the design here and the separation of responsibilities?

One good thin to know would be: say we had a bug in hitting 'enter'.  How would you determine which of these three systems was responsible for handling fixing it?

heejaechang commented:
we have 3 different scenario.

1. format - people issue format command
2. indentation - people ask where caret should go on blank line
3. auto formatting - this do bunch stuff
a. if it is one of certain token like ";" or "}", it finds starting point (which is heuristic on what considered as starting point) and format those span.
b. if it is enter, then it sees whether caret was before certain token like { or whether it is at the start location of token, and whether it is enter on a whitespace, and do things differently. here it is full case by case people added as people get feedbacks.

also, for 2 and 3, since formatting doesn't touch any expression level construct except certain cases like query expression, for 2 and 3 to handle "enter" or "indentation" of expression level, it adds some extra rules for things like argument list, parameter list and etc to format engine.

also for us to not re-implementation logic to determine where to put caret based on various formatting options for indentation, it uses formatting rules to determine indentation if there is new line indentation rule. otherwise, it uses heuristic to find indentation since there is no option to change its behavior.

and there are bunch of case by case for indentation from feedbacks. can't remember all those special cases.

heejaechang commented:
ah, and we also has special case for razor and venus through some formatting rule provider and etc. we can't get rid of that due to venus. but hopefully for new razor, we don't need to use it.

CyrusNajmabadi commented:
Could you clarify what the intended split is across all three indentation systems today?

CyrusNajmabadi commented:
> we have 3 different scenario.

Ok.  But hte only scenario i'm really asking about is 'indentation/enter-handling'.  My question is about why we've separated the logic out for that into mulitple different locations, with each location needing to know about and poke into the other locations to figure out what's going on.  That's the part that seems very confusing, and makes it very difficult for, say, a feature to ask "where should i indent here?"

CyrusNajmabadi commented:
Let me try to ask teh question as simply as possible:

We have: 

```c#
internal partial class SmartIndent : ISmartIndent
```

And

```c#
internal class SmartTokenFormatterCommandHandler
```

Both are responsible for figuring out where to place the care after `enter` is hit.  My questions are these:

1. Why do we have both of these?
2. What determines which logic each is responsible for? 
3. Why can't we just merge the logic of these into the other?  

I can't figure out any rhyme or reasons on these guys.

--

Note: if gitter would be better to discuss things, i'm happy to take it there.  It might be easier to ask questions and build understanding versus trying to do this back and forth over github.  Which would you prefer?  Thanks!

CyrusNajmabadi commented:
> I guess it is time to clean things up here.

Alright, i'll see what i can do.

CyrusNajmabadi commented:
> I guess it is time to clean things up here.

@heejaechang i have cleaned up a lot here: https://github.com/dotnet/roslyn/pull/33989

This removes unnecessary separation of services for handling indentation.

--

THe end result gets rid of a lot.  but it still ends up being complex.  That's because the indentation service itself has all these different codepaths trying to decide what behavior to perform with all sorts of complex assumptions. I have no idea  how to make that better, but at least it's all contained in this one service now.

CyrusNajmabadi commented:
> Ensuring this gets a design review to determine the approach we want for two specific items:

> Are type forwarders preferred to redefining types?

I would prefer not here.  There are changes i want to make to the interfaces in other PRs (for example, making the return types non-nullable).  So this will be more than just a move.

CyrusNajmabadi commented:
> We have active plans to eliminate external code depending on these interfaces, and need to determine which we want to pursue first.

Can you explain these plans?  What's the recommended way for a 3rd party language to participate in indentation?

sharwell commented:
> Can you explain these plans? 

All components not in dotnet/roslyn (except for VS for Mac) must now go through an adapter assembly rather than directly access internal Roslyn APIs.

#31393

> What's the recommended way for a 3rd party language to participate in indentation?

Through the public API.

sharwell commented:
I would prefer to migrate existing external components off of the internal indentation APIs *before* making changes that go beyond the abilities of type forwarders. Will need to review with F# and TypeScript to determine the complexity of these changes.

CyrusNajmabadi commented:
What public API are you referring to?

CyrusNajmabadi commented:
> I would prefer to migrate existing external components off of the internal indentation APIs before making changes that go beyond the abilities of type forwarders. Will need to review with F# and TypeScript to determine the complexity of these changes. Otherwise this change is vastly more complex to account for a situation that we can avoid from the start.

The change at the API level here is super tiny.  I think it's pretty clear that any Roslyn language would be able to move from the older API to this one. And it's not like the old API goes away. 

Can we not add this, expose the public endpoint and move the other languages over.  

I worry about this process taking months of waiting (for something simple), instead of just easily happening

sharwell commented:
> What public API are you referring to?

If there is no public API for the feature, then each external consumer would be expected to implement and maintain their own API surface within their respective ExternalAccess assembly. This pattern/API would be adjusted as necessary to best meet the needs of that teem.

CyrusNajmabadi commented:
> If there is no public API for the feature, then each external consumer would be expected to implement and maintain their own API surface within their respective ExternalAccess assembly. This pattern/API would be adjusted as necessary to best meet the needs of that teem.

I'm really confused.  This is like *the* most simple IDE feature to integrate with other languages with.  Why would this not just be as simple as the signature we now have?

Can you roll the right people into this discussion so that we can make forward progress on this?  I really don't want this to stagnate.  Everything is fresh in my head, and if this takes months like normal, it's just going to be so much more difficult to handle when no one has been thinking about this for ages.

CyrusNajmabadi commented:
I'm not really sure how this is intended to work.  I took a look at the PR and the projects checked in, and I can't see how it actually does anything:

![image](https://user-images.githubusercontent.com/4564579/54143038-31e2d880-43e6-11e9-9f69-11b9c5d5d669.png)

It seems to be entirely empty afaict.

sharwell commented:
> Can you roll the right people into this discussion so that we can make forward progress on this?

It's on the design review agenda for today.

sharwell commented:
> I'm not really sure how this is intended to work. I took a look at the PR and the projects checked in, and I can't see how it actually does anything:

The sequence is roughly:

1. Create adapter assemblies where IVT consumers can define stable APIs
2. Update IVT consumers to define APIs in the adapter assemblies, and use those instead of direct access to APIs elsewhere
3. Remove external IVT access everywhere except the adapter assemblies
4. Refactor code in Roslyn for clarity, which is safe and clean now that external implementations cannot reference these internal implementation details

This pull request would move step 4 before step 2, which I find impairs clarity of the work towards the final goal.

CyrusNajmabadi commented:
> Create adapter assemblies where IVT consumers can define stable APIs
Update IVT consumers to define APIs in the adapter assemblies, and use those instead of direct access to APIs elsewhere

How do we get '1' and '2' to happen (especially in any sort of reasonable timeframe)?  Thanks!

CyrusNajmabadi commented:
> It's on the design review agenda for today.

Ok, thanks!

sharwell commented:
> How do we get '1' and '2' to happen (especially in any sort of reasonable timeframe)? Thanks!

Large scale refactorings are excruciatingly slow in Roslyn due to the large number of IVTs that came in over time. We now have a plan to remove them via one final refactoring, which is step '2'. Like the ones that preceded it, I expect this to be excruciatingly slow. But after that, we can start to pay down some of the other debt that accumulated over time. In the meantime, I expect PRs like #33722 (and possibly this one), to be delayed.

CyrusNajmabadi commented:
I don't see why any of this work needs to be slowed down.  The existing interfaces are still here and still work fine.  We can still do all the debt paydown work.  The final bit can all be all this stuff you're mentioning (i.e. the IVT part and the 3rd party languages moving over).

All that is orthogonal work and can be done independently.  It seems strange to gate *this* work on that other stuff.  How is that beneficial?

heejaechang commented:
I don't know why things are splited. like I said, formatting handler used to only handle format document/selection command and smart token hander handled everything else. not sure they things got splited the way it is now.

indentation service (http://index/?query=ISmartI&rightProject=Microsoft.VisualStudio.Text.UI&file=Editor%5CISmartIndentProvider.cs&line=18) is Editor service that we need to implement that has nothing to do with command handler. and I believe there is some case where we need to decide whether we should handle the case through command handler or we just let editor to handle the case through the indentation service.

why we behave differently in certain case by us handling enter ourselves rather than let indentation service that care of it? I dont know, I wasn't one that decide that behavior. but as long as I know current behavior follows behavior since like Orcas. so there should be good reason to change the behavior if we want to change.

CyrusNajmabadi commented:
> did you check history

Yes.  The history shows this strange split between SmartIndent and the SmartTokenCommandHandler as far back as the beginning.  I honestly don't know why we have this split.  So i removed it and cleaned up a lot of stuff through the following PRs:

https://github.com/dotnet/roslyn/pull/33978
https://github.com/dotnet/roslyn/pull/33983
https://github.com/dotnet/roslyn/pull/33989
https://github.com/dotnet/roslyn/pull/34001

CyrusNajmabadi commented:
> and smart token hander handled everything else

This is not the case.  There is the SmartIndent code, and there is the SmartTokenCommandHandler code.  As far as i can tell, these have been around since the beginning.

For example, here is the first Roslyn open PR containing SmartIndent:

https://github.com/dotnet/roslyn/blob/4754991da6685dfe4d7fcb967e8163490d12df8a/src/EditorFeatures/Core/Implementation/SmartIndent/SmartIndent.cs

At that same time, we also had the SmartTokenCommandHandler: 
https://github.com/dotnet/roslyn/blob/4754991da6685dfe4d7fcb967e8163490d12df8a/src/EditorFeatures/Core/Implementation/Formatting/Indentation/AbstractSmartTokenFormatterCommandHandler.cs

These types seem to be utterly redundant.  And i cannot fathom the purpose of the latter given that the former exists.

CyrusNajmabadi commented:
@heejaechang I cleaned things up a bunch here.  First, i got rid of what seems to eb a totally useless formatting service in https://github.com/dotnet/roslyn/pull/33983.  Basically, we had an entire service/handler to handle someone typing `<enter>` after `using (foo)`.  This basically doesn't behave at all like any of the rest of the formatting/indentation code for the rest of C#, so i just torched it.

Second, i got rid of hte nonsensical split between SmartIndent and SmartTokenCommandHandler.  I cannot understnad why we have both of these.  Furthermore, SmartIndent used to have all this awkward code in it saying "let me check if SmartTokenCommandHandler will run.  if so, i'll diable myself". 

This whole system was brittle and confusing and nonsensical.

--

The end result is much better than before, but it could still use a lot of cleanup.  Specifically, the actual remaining indentation code (which is now localized to SmartIndent) is extremely complex and confusing.  It seems to just be a mess of cases and random logic paths, and it's really hard to actually understand why the code does what it does, or why some pieces of code are in some places and not the other.  It should really be cleaned up in the future, but at least it's all localized within teh SmartIndent system, and not scattered across many different disparate feature locations.

CyrusNajmabadi commented:
@sharwell did you make a design decision on this from monday?

CyrusNajmabadi commented:
closing

sharwell commented:
Design review conclusion: we can take this for "preview 1" step (16.1.P1 or 16.2.P1), depending on when the reviews are completed. We don't have a date for when TypeScript will complete the move to ExternalAccess so it's not blocked on that.

CyrusNajmabadi commented:
@sharwell Regarding:

> Design review conclusion: we can take this for "preview 1" step (16.1.P1 or 16.2.P1), depending on when the reviews are completed. We don't have a date for when TypeScript will complete the move to ExternalAccess so it's not blocked on that.

What are the next steps here?  Can you/@heejaechang review when convenient?  Note that this is one of three PRs intended to go in a row:

- [ ] This one.
- [ ] https://github.com/dotnet/roslyn/pull/33989. "Merge all indentation related code into one service."
- [ ] https://github.com/dotnet/roslyn/pull/34001. "Remove IBlankLineIndentationService".

So this PR is not intended to be hte endpoint.  But just a piece that could break out nicely to make the review easier to handle.  Thanks!

CyrusNajmabadi commented:
ping @sharwell .  thanks!

CyrusNajmabadi commented:
ping @sharwell @heejaechang . thanks!

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Indentation/IIndentationService.cs:
```diff
@@ -0,0 +1,69 @@
+        /// <see langword="null"/> if the line in question is not blank and thus indentation should
+        /// be deferred to the formatting command handler to handle.
+        /// </summary>
+        IndentationResult? GetDesiredIndentation(Document document, int lineNumber, CancellationToken cancellationToken);
```
Note: a later PR removes the lower method, and makes this method non-optional.  This PR is just a stepping stone to that.

CyrusNajmabadi replied:
The final shape of this interface is the simple and clear:

```c#
    internal interface IIndentationService : ILanguageService
    {
        /// <summary>
        /// Determines the desired indentation of a given line.
        /// </summary>
        IndentationResult GetIndentation(
            Document document, int lineNumber,
            FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken);
    }
```

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Indentation/IIndentationService.cs:
```diff
@@ -0,0 +1,69 @@
+        /// This function will always succeed.
+        /// </summary>
+        IndentationResult GetBlankLineIndentation(
+            Document document, int lineNumber, FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken);
```
This method is removed in: https://github.com/dotnet/roslyn/pull/34001.  it's here only to help enable tehse individual refactoring steps.

CyrusNajmabadi commented:
@heejaechang can you ptal.  Thanks!

CyrusNajmabadi commented:
@heejaechang can you ptal. Thanks!

CyrusNajmabadi commented:
@jinujoseph Is there someone who can help with this PR?

CyrusNajmabadi commented:
@dpoeschl or @jasonmalinowski could you maybe look at this?

sharwell commented:
Moving back to blocked on the second part of https://github.com/dotnet/roslyn/pull/33978#pullrequestreview-212570114. This concern has since changed:

> We don't have a date for when TypeScript will complete the move to ExternalAccess so it's not blocked on that.

The review for this refactoring was opportunistic since we didn't when the External Access layer would be ready to accommodate these refactorings. However, we were not able to review this change yet and we are now planning to have the External Access layer transition complete for 16.2 Preview 1. Refactorings reviewed and merged after this transition will be more efficient than refactorings performed before this change, and are also less likely to interfere with that work going on in parallel.

CyrusNajmabadi commented:
@sharwell Can we move forward on this?

CyrusNajmabadi commented:
@sharwell @jinujoseph  can we move forward on this?

tmat commented on src/EditorFeatures/Core/Implementation/Formatting/Indentation/AbstractSmartTokenFormatterCommandHandler.cs:
```diff
@@ -139,7 +140,7 @@ internal void ExecuteCommandWorker(ReturnKeyCommandArgs args, CancellationToken
             }
 
-            var indentationService = document.GetLanguageService<ISynchronousIndentationService>();
+            var indentationService = document.GetLanguageService<CodeAnalysis.Indentation.IIndentationService>();
```
>CodeAnalysis.Indentation [](start = 65, length = 24)

Is this qualification needed?

CyrusNajmabadi replied:
fixed.

tmat commented on src/EditorFeatures/Core/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:
```diff
@@ -17,7 +16,7 @@ internal abstract partial class AbstractBinaryExpressionWrapper<TBinaryExpressio
 
         protected AbstractBinaryExpressionWrapper(
-            IBlankLineIndentationService indentationService,
+            Indentation.IIndentationService indentationService,
```
>Indentation [](start = 12, length = 11)

Let's add a using instead of qualifying.

CyrusNajmabadi replied:
sure.

CyrusNajmabadi replied:
Done.

tmat commented:
@CyrusNajmabadi Can you file an issue tracking removal of the newly obsolete API and link it from the code?

CyrusNajmabadi commented:
> @CyrusNajmabadi Can you file an issue tracking removal of the newly obsolete API and link it from the code?

@tmat Done!

CyrusNajmabadi commented:
@tmat feel free to merge when you havea  chance.

tmat commented:
@heejaechang @sharwell OK to merge?

CyrusNajmabadi commented:
> @heejaechang @sharwell OK to merge?

Ping.  Thanks!

CyrusNajmabadi commented:
@tmat, @sharwell says he's not blocking this, but he's not moving it forward.  Would it be possible to directly pull in @heejaechang as i find pings don't reach him very well.

jinujoseph commented:
spoke to @heejaechang , he will take a look at it soon.

heejaechang commented on src/EditorFeatures/CSharp/SplitStringLiteral/SplitStringLiteralCommandHandler.StringSplitter.cs:
```diff
@@ -10,6 +9,8 @@
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.SplitStringLiteral
 {
+    using Microsoft.CodeAnalysis.Indentation;
```
move to where other using is?

CyrusNajmabadi replied:
can't.  if we move it, then diffrent symbols will be found in Microsoft.CodeAnalysis.Editor.

heejaechang replied:
so, we have a plan to move at some point then? or it has to be here until typescript/F# is moving to new API as well?

CyrusNajmabadi replied:
Yes, this is tracked with: https://github.com/dotnet/roslyn/issues/35872

heejaechang commented on src/EditorFeatures/Core/Implementation/Formatting/Indentation/AbstractSmartTokenFormatterCommandHandler.cs:
```diff
@@ -23,6 +23,8 @@
 
 namespace Microsoft.CodeAnalysis.Editor.Implementation.Formatting.Indentation
 {
+    using Microsoft.CodeAnalysis.Indentation;
```
same here?

CyrusNajmabadi replied:
can't.  if we move it, then diffrent symbols will be found in Microsoft.CodeAnalysis.Editor.

heejaechang commented on src/EditorFeatures/Core/Implementation/SmartIndent/SmartIndent.cs:
```diff
@@ -40,21 +44,39 @@ public void Dispose()
-                if (asyncService != null)
+                // If we don't have a feature-layer service, try to fall back to the legacy
+                // editor-feature-layer interfaces.
+
```
is this for things like typescript or F#?

CyrusNajmabadi replied:
Yup!

heejaechang commented on src/EditorFeatures/Core/Shared/Options/FeatureOnOffOptions.cs:
```diff
@@ -44,10 +44,6 @@ internal static class FeatureOnOffOptions
             nameof(FeatureOnOffOptions), nameof(AutoFormattingOnTyping), defaultValue: true,
             storageLocations: new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.Auto Formatting On Typing"));
 
-        public static readonly PerLanguageOption<bool> AutoFormattingOnReturn = new PerLanguageOption<bool>(
```
where it go?

heejaechang commented on src/EditorFeatures/Core/Wrapping/BinaryExpression/AbstractBinaryExpressionWrapper.cs:
```diff
@@ -6,18 +6,19 @@
 
 namespace Microsoft.CodeAnalysis.Editor.Wrapping.BinaryExpression
 {
+    using Microsoft.CodeAnalysis.Indentation;
```
same here.

CyrusNajmabadi replied:
can't.  if we move it, then different symbols will be found in Microsoft.CodeAnalysis.Editor.

heejaechang commented on src/EditorFeatures/Core/Wrapping/SeparatedSyntaxList/AbstractSeparatedSyntaxListWrapper.cs:
```diff
@@ -5,6 +5,8 @@
 
 namespace Microsoft.CodeAnalysis.Editor.Wrapping.SeparatedSyntaxList
 {
+    using Microsoft.CodeAnalysis.Indentation;
```
here.

CyrusNajmabadi replied:
can't.  if we move it, then different symbols will be found in Microsoft.CodeAnalysis.Editor.

heejaechang commented on src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs:
```diff
@@ -99,6 +99,9 @@ private static string GetEndOfLineEditorConfigString(string option)
 
         internal static Option<bool> AllowDisjointSpanMerging { get; } = CreateOption(OptionGroup.Default, nameof(AllowDisjointSpanMerging), defaultValue: false);
 
+        internal static readonly PerLanguageOption<bool> AutoFormattingOnReturn = CreatePerLanguageOption(OptionGroup.Default, nameof(AutoFormattingOnReturn), defaultValue: true,
```
ah. I see. it moved to here.

heejaechang commented on src/Workspaces/Core/Portable/Formatting/FormattingOptions.cs:
```diff
@@ -99,6 +99,9 @@ private static string GetEndOfLineEditorConfigString(string option)
         internal static Option<bool> AllowDisjointSpanMerging { get; } = CreateOption(OptionGroup.Default, nameof(AllowDisjointSpanMerging), defaultValue: false);
 
+        internal static readonly PerLanguageOption<bool> AutoFormattingOnReturn = CreatePerLanguageOption(OptionGroup.Default, nameof(AutoFormattingOnReturn), defaultValue: true,
+            storageLocations: new RoamingProfileStorageLocation("TextEditor.%LANGUAGE%.Specific.Auto Formatting On Return"));
```
this option location is very specific to VS. so I am not sure what it means that Workspace layer saving option to place where it is specific to VS.

CyrusNajmabadi replied:
the option itself is a bit strange.  it's used in one place:

```c#
        public static bool ShouldUseSmartTokenFormatterInsteadOfIndenter(
            IEnumerable<AbstractFormattingRule> formattingRules,
            CompilationUnitSyntax root,
            TextLine line,
            OptionSet optionSet,
            CancellationToken cancellationToken)
        {
            Contract.ThrowIfNull(formattingRules);
            Contract.ThrowIfNull(root);

            if (!optionSet.GetOption(FormattingOptions.AutoFormattingOnReturn, LanguageNames.CSharp))
            {
                return false;
            }
```

Which is in the indentation service itself.  I think it's a way to allow us (potentially for manual testing purposes) to opt out of smart-indent.  It's a bit wonky, but i'm not sure to do about it.  We could try to figure something else out once the following two PRs go in here.

CyrusNajmabadi commented:
@heejaechang if you want, feel free to merge in.  I'll move ahead with teh followup PRs and we can then decide how best to structure the options here.

CyrusNajmabadi commented:
@heejaechang can you please merge in?

CyrusNajmabadi commented:
ping @heejaechang

CyrusNajmabadi commented:
Thanks!

CyrusNajmabadi commented:
@heejaechang Did you squash this PR?  I've having a heck of a time trying to merge it into all my other branches :-/
